### PR TITLE
Support pyspark.sql.Dataframe in TFDataset

### LINKED ISSUE #1898: tf.keras model fit couldn't support DataFrame interface
Most of time we already used DataFrame to do feature engineering but we have to call  TFDataset.from_rdd interface to transfer the DF to TFDataSet interface and there is no "from_dataframe" function from TFDataSet. And the developer has to go to low level API to transfer rdd to data structures , such as call lambda , np.array etc..
   Will be better experience to support high level Spark data frame interface and simply the function parameters.
       
    trainingDF = genData(tDF, "201312", "201401")
    validationDF = genData(tDF, "201402", "201403")
    model = getKerasModel()
    
    train_rdd = trainingDF.rdd.map(lambda row: (np.array(row["features"]), np.array(int(row["labels"]))))
    validation_rdd = validationDF.rdd.map(lambda row: (np.array(row["features"]), np.array(int(row["labels"]))))
    dataset = TFDataset.from_rdd(train_rdd, features=(tf.float32, [3]), labels=(tf.int32, []), batch_size=20,val_rdd=validation_rdd)
    
    model.fit(dataset,max_epoch,distributed=True)

### PR DESCRIPTION
**Pull request by yangw1234**
Fix #1898 
Usage
```python
rdd = sc.range(0, 1000)
df = rdd.map(lambda x: (DenseVector(np.random.rand(20).astype(np.float)),
                                x % 10)).toDF(["feature", "label"])
train_df, val_df = df.randomSplit([0.7, 0.3])
dataset = TFDataset.from_dataframe(train_df,
                         feature_cols=["feature"],
                         labels_cols=["label"],
                         batch_size=32,
                         validation_df=val_df)
model = KerasModel(get_keras())
model.fit(dataset)
```

Limitations:
1. only support scalar or vector types
2. only support numeric types

### CODE REVIEW

**Comment by yangw1234** [comment]
jenkins pass http://172.168.2.101:8080/view/ZOO-PR/job/ZOO-PR-Validation/2980/

File: docs/docs/APIGuide/TFPark/tf-dataset.md [hunk 1]
```diff
@@ -263,4 +263,27 @@ from_tf_data_dataset(dataset, batch_size=-1, batch_per_thread=-1, hard_code_batc
...
+
+**Python**
+```python
+from_dataframe(df, feature_cols, labels_cols=None, batch_size=-1, batch_per_thread=-1, hard_code_batch_size=False, validation_df=None)
+```
+
+**Arguments**
+
+* **df**: the DataFrame for the dataset
+* **feature_cols**: a list of string, indicating which columns are used as features
```
Comment (jason-dai) [inline docs/docs/APIGuide/TFPark/tf-dataset.md]: what are the types of the feature and label columns supported?

File: pyzoo/zoo/tfpark/tf_dataset.py [hunk 2]
```diff
@@ -1098,3 +1123,103 @@ def from_ndarrays(tensors, batch_size=-1, batch_per_thread=-1,
...
+        if isinstance(dtype, df_types.DoubleType):
+            return True
+        return False
+
+    def __init__(self, df, feature_cols, labels_cols=None, batch_size=-1,
+                 batch_per_thread=-1, hard_code_batch_size=False,
+                 validation_df=None,
+                 sequential_order=False, shuffle=True):
+        assert isinstance(feature_cols, list), "feature_cols should be a list"
+        assert isinstance(labels_cols, list), "label_cols should be a list"
```
Comment (jason-dai) [inline pyzoo/zoo/tfpark/tf_dataset.py]: is None a list?

**Comment by jason-dai** [comment]
Any unit test added?

File: pyzoo/zoo/tfpark/tf_dataset.py [hunk 3]
```diff
@@ -1098,3 +1123,103 @@ def from_ndarrays(tensors, batch_size=-1, batch_per_thread=-1,
...
+            tf_type, tf_shape = DataFrameDataset.df_datatype_to_tf(data_type)
+            feature_meta.append(TensorMeta(tf_type, name=name, shape=tf_shape))
+
+        if labels_cols:
+            label_meta = []
+            for label_col in labels_cols:
+                field = schema[label_col]
+                name = field.name
+                data_type = field.dataType
+                if DataFrameDataset.df_datatype_to_tf(data_type) is None:
```
Comment (jason-dai) [inline pyzoo/zoo/tfpark/tf_dataset.py]: why only check this for label, not feature?

**Comment by yangw1234** [comment]
@jason-dai all comments addressed in PR #1987 and improved test coverage

**Comment by jack1981** [comment]
Got exception when run this function 
   return f(*args, **kwargs)
  File "/dev/shm/zoo/lib/analytics-zoo-bigdl_0.10.0-spark_2.4.3-0.9.0-SNAPSHOT-python-api.zip/zoo/tfpark/tf_dataset.py", line 1174, in <lambda>
  File "/dev/shm/zoo/lib/analytics-zoo-bigdl_0.10.0-spark_2.4.3-0.9.0-SNAPSHOT-python-api.zip/zoo/tfpark/tf_dataset.py", line 1167, in convert
  File "/dev/shm/zoo/lib/analytics-zoo-bigdl_0.10.0-spark_2.4.3-0.9.0-SNAPSHOT-python-api.zip/zoo/tfpark/tf_dataset.py", line 1162, in convert_for_cols
AssertionError: unsupported field StructField(features,ArrayType(DoubleType,false),true), data [3706.0, 1.0, 1.0]

        at org.apache.spark.api.python.BasePythonRunner$ReaderIterator.handlePythonException(PythonRunner.scala:452)
        at org.apache.spark.api.python.PythonRunner$$anon$1.read(PythonRunner.scala:588)
        at org.apache.spark.api.python.PythonRunner$$anon$1.read(PythonRunner.scala:571)
        at org.apache.spark.api.python.BasePythonRunner$ReaderIterator.hasNext(PythonRunner.scala:406)
        at org.apache.spark.InterruptibleIterator.hasNext(InterruptibleIterator.scala:37)
        at scala.collection.Iterator$$anon$12.hasNext(Iterator.scala:440)
        at scala.collection.Iterator$$anon$11.hasNext(Iterator.scala:409)
        at scala.collection.Iterator$$anon$11.hasNext(Iterator.scala:409)
        at org.apache.spark.shuffle.sort.BypassMergeSortShuffleWriter.write(BypassMergeSortShuffleWriter.java:125)
        at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:99)
        at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:55)
        at org.apache.spark.scheduler.Task.run(Task.scala:121)
        at org.apache.spark.executor.Executor$TaskRunner$$anonfun$10.apply(Executor.scala:408)
        at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1405)
        at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:414)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)

**Comment by jack1981** [comment]
Example code as follow: 
TFDataset.from_rdd works but TFDataset.from_dataframe did not work !

trainingDF = genData(tDF, train_start, train_end)
    validationDF = genData(tDF, validation_start, validation_end)
    testDF = genData(tDF, test_start,test_end)
    model = getKerasModel(log_hdfs_dir)    
    train_rdd = trainingDF.rdd.map(lambda row: (np.array(row["features"]), np.array(int(row["labels"]))))
    validation_rdd = validationDF.rdd.map(lambda row: (np.array(row["features"]), np.array(int(row["labels"]))))
    test_rdd = testDF.rdd.map(lambda row: (np.array(row["features"]), np.array(int(row["labels"]))))
    # each element in train_rdd should be a tuple of two, all of which should be numpy.ndarrays,
    #dataset = TFDataset.from_dataframe(trainingDF,feature_cols=["features"],labels_cols=["labels"],batch_size=batch_size,validation_df=validationDF)
    dataset = TFDataset.from_rdd(train_rdd, features=(tf.float32, [3]), labels=(tf.int32, []), batch_size=batch_size,val_rdd=validation_rdd)
    # kick off training
    model.fit(dataset,epochs=max_epoch,distributed=True)
    # Evaluation with test datasets
    eval_dataset = TFDataset.from_rdd(test_rdd, features=(tf.float32, [3]), labels=(tf.int32, []), batch_size=batch_size)
    #eval_dataset = TFDataset.from_dataframe(testDF,feature_cols=["features"],labels_cols=["labels"],batch_size=batch_size)

**Comment by hkvision** [comment]
@yangw1234 Please take a look at this.

**Comment by yangw1234** [comment]
@jack1981 @hkvision this should have been fixed by #2264

**Comment by jack1981** [comment]
It worked after we got the night build !

But we still had some exceptions ( not sure relevant to this issue)  came out.
Looks like if add model_dir parameter , after every Epoch , zoo will save model and metrics ? 
  keras_model = KerasModel(model,model_dir=hdfs_path)

If the model is big , we had 2 G model size , do you think it will be an issue for HDFS I/O operation ? 

2020-04-26 21:27:20 INFO  DistriOptimizer$:180 - [Epoch 1 62400/61289][Iteration 39][Wall Clock 9.266737315s] bigdl_metirc_0 Loss is (Loss: 144.88086, count: 224, Average Loss: 0.64678955)
2020-04-26 21:27:20 INFO  DistriOptimizer$:180 - [Epoch 1 62400/61289][Iteration 39][Wall Clock 9.266737315s] bigdl_metirc_1 Top1Accuracy is Accuracy(correct: 6672, count: 10181, accuracy: 0.6553383754051665)
2020-04-26 21:27:21 INFO  DistriOptimizer$:220 - [Wall Clock 9.306474856s] Save model to hdfs://dw-prod-new/user/dw/zoo/logs/2020-04-26_21-25-47/models/20200426_212710
2020-04-26 21:27:21 INFO  DistriOptimizer$:225 - [Wall Clock 9.306474856s] Save optimMethod com.intel.analytics.bigdl.optim.RMSprop@7b88934b to hdfs://dw-prod-new/user/dw/zoo/logs/2020-04-26_21-25-47/models/20200426_212710
2020-04-26 21:27:21 INFO  DistriOptimizer$:413 - [Epoch 2 1600/61289][Iteration 40][Wall Clock 9.401520679s] Trained 1600 records in 0.095045823 seconds. Throughput is 16833.986 records/second. Loss is 0.6227355.
2020-04-26 21:27:21 ERROR ThreadPool$:224 - Error: java.nio.channels.ClosedChannelException
        at org.apache.hadoop.hdfs.DFSOutputStream.checkClosed(DFSOutputStream.java:2039)
        at org.apache.hadoop.fs.FSOutputSummer.write(FSOutputSummer.java:105)
        at org.apache.hadoop.fs.FSDataOutputStream$PositionCache.write(FSDataOutputStream.java:58)
        at java.io.DataOutputStream.write(DataOutputStream.java:107)
        at java.io.FilterOutputStream.write(FilterOutputStream.java:97)
        at com.intel.analytics.bigdl.visualization.tensorboard.RecordWriter.write(RecordWriter.scala:44)
        at com.intel.analytics.bigdl.visualization.tensorboard.EventWriter.writeEvent(EventWriter.scala:57)
        at com.intel.analytics.bigdl.visualization.tensorboard.EventWriter.run(EventWriter.scala:68)
        at com.intel.analytics.bigdl.visualization.tensorboard.FileWriter$$anonfun$1.apply$mcV$sp(FileWriter.scala:40)
        at com.intel.analytics.bigdl.visualization.tensorboard.FileWriter$$anonfun$1.apply(FileWriter.scala:40)
        at com.intel.analytics.bigdl.visualization.tensorboard.FileWriter$$anonfun$1.apply(FileWriter.scala:40)
        at com.intel.analytics.bigdl.utils.ThreadPool$$anonfun$invoke$2.apply(ThreadPool.scala:221)
        at scala.concurrent.impl.Future$PromiseCompletingRunnable.liftedTree1$1(Future.scala:24)
        at scala.concurrent.impl.Future$PromiseCompletingRunnable.run(Future.scala:24)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
        at java.lang.Thread.run(Thread.java:748)

2020-04-26 21:27:21 INFO  DistriOptimizer$:413 - [Epoch 2 3200/61289][Iteration 41][Wall Clock 9.494740924s] Trained 1600 records in 0.093220245 seconds. Throughput is 17163.654 records/second. Loss is 0.6436238.

**Comment by yangw1234** [comment]
hi @jack1981 this problem is unrelated to TFDataset, I have created another issue #2269 to track this.

**Comment by jack1981** [comment]
When we run yarn client mode , it can finish but had previous exception 
But when we run yarn cluster mode, we got another exception, here are the details : 

+------+----+------+-----+-------------------+
|   uid| mid|labels|count|           features|
+------+----+------+-----+-------------------+
| 344.0|21.0|   0.0|    2| [344.0, 21.0, 2.0]|
| 238.0|98.0|   0.0|    0| [238.0, 98.0, 0.0]|
| 785.0| 2.0|   0.0|    0|  [785.0, 2.0, 0.0]|
|3318.0|23.0|   0.0|    0|[3318.0, 23.0, 0.0]|
|2196.0|12.0|   0.0|    1|[2196.0, 12.0, 1.0]|
+------+----+------+-----+-------------------+
only showing top 5 rows

pos:  6271 neg:  3914
label df count:  10185
+------+----+------+-----+-------------------+
|   uid| mid|labels|count|           features|
+------+----+------+-----+-------------------+
|1574.0| 4.0|   0.0|    2| [1574.0, 4.0, 2.0]|
|   4.0| 9.0|   0.0|   27|   [4.0, 9.0, 27.0]|
|1966.0|11.0|   0.0|    0|[1966.0, 11.0, 0.0]|
| 457.0| 3.0|   0.0|    7|  [457.0, 3.0, 7.0]|
| 527.0| 6.0|   0.0|    3|  [527.0, 6.0, 3.0]|
+------+----+------+-----+-------------------+
only showing top 5 rows

pos:  6272 neg:  3918
label df count:  10190
+------+----+------+-----+-------------------+
|   uid| mid|labels|count|           features|
+------+----+------+-----+-------------------+
|   4.0| 9.0|   0.0|   35|   [4.0, 9.0, 35.0]|
|2041.0| 2.0|   0.0|    1| [2041.0, 2.0, 1.0]|
|3678.0|11.0|   0.0|    0|[3678.0, 11.0, 0.0]|
| 853.0| 6.0|   0.0|    1|  [853.0, 6.0, 1.0]|
|3176.0|89.0|   0.0|    1|[3176.0, 89.0, 1.0]|
+------+----+------+-----+-------------------+
only showing top 5 rows

WARNING:tensorflow:From /dfs/1/yarn/nm/usercache/dw/appcache/application_1586627890098_48586/container_e21_1586627890098_48586_02_000001/venv.zip/venv/lib/python3.6/site-packages/tensorflow_core/python/keras/initializers.py:119: calling RandomUniform.__init__ (from tensorflow.python.ops.init_ops) with dtype is deprecated and will be removed in a future version.
Instructions for updating:
Call initializer instance with the dtype argument instead of passing it to the constructor
WARNING:tensorflow:OMP_NUM_THREADS is no longer used by the default Keras config. To configure the number of threads, use tf.config.threading APIs.
2020-04-26 21:47:12.901176: W tensorflow/stream_executor/platform/default/dso_loader.cc:55] Could not load dynamic library 'libcuda.so.1'; dlerror: libcuda.so.1: cannot open shared object file: No such file or directory
2020-04-26 21:47:12.901217: E tensorflow/stream_executor/cuda/cuda_driver.cc:318] failed call to cuInit: UNKNOWN ERROR (303)
2020-04-26 21:47:12.901245: I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:156] kernel driver does not appear to be running on this host (hdp2stl020005.mastercard.int): /proc/driver/nvidia/version does not exist
2020-04-26 21:47:12.901617: I tensorflow/core/platform/cpu_feature_guard.cc:142] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2 AVX512F FMA
2020-04-26 21:47:12.917515: I tensorflow/core/platform/profile_utils/cpu_utils.cc:94] CPU Frequency: 2100000000 Hz
2020-04-26 21:47:12.920890: I tensorflow/compiler/xla/service/service.cc:168] XLA service 0x55baf2a8dac0 initialized for platform Host (this does not guarantee that XLA will be used). Devices:
2020-04-26 21:47:12.920915: I tensorflow/compiler/xla/service/service.cc:176]   StreamExecutor device (0): Host, Default Version
WARNING:tensorflow:From /dfs/1/yarn/nm/usercache/dw/appcache/application_1586627890098_48586/container_e21_1586627890098_48586_02_000001/venv.zip/venv/lib/python3.6/site-packages/tensorflow_core/python/ops/resource_variable_ops.py:1630: calling BaseResourceVariable.__init__ (from tensorflow.python.ops.resource_variable_ops) with constraint is deprecated and will be removed in a future version.
Instructions for updating:
If using Keras pass *_constraint arguments to layers.
Model: "model"
__________________________________________________________________________________________________
Layer (type)                    Output Shape         Param #     Connected to                     
==================================================================================================
input_1 (InputLayer)            [(None, 3)]          0                                            
__________________________________________________________________________________________________
tf_op_layer_strided_slice (Tens [(None,)]            0           input_1[0][0]                    
__________________________________________________________________________________________________
tf_op_layer_strided_slice_1 (Te [(None,)]            0           input_1[0][0]                    
__________________________________________________________________________________________________
embedding (Embedding)           (None, 100)          2000100     tf_op_layer_strided_slice[0][0]  
__________________________________________________________________________________________________
embedding_1 (Embedding)         (None, 50)           10050       tf_op_layer_strided_slice_1[0][0]
__________________________________________________________________________________________________
flatten (Flatten)               (None, 100)          0           embedding[0][0]                  
__________________________________________________________________________________________________
flatten_1 (Flatten)             (None, 50)           0           embedding_1[0][0]                
__________________________________________________________________________________________________
concatenate (Concatenate)       (None, 150)          0           flatten[0][0]                    
                                                                 flatten_1[0][0]                  
__________________________________________________________________________________________________
dense (Dense)                   (None, 75)           11325       concatenate[0][0]                
__________________________________________________________________________________________________
dense_1 (Dense)                 (None, 2)            152         dense[0][0]                      
==================================================================================================
Total params: 2,021,627
Trainable params: 2,021,627
Non-trainable params: 0
__________________________________________________________________________________________________
WARNING:tensorflow:From /dfs/1/yarn/nm/usercache/dw/appcache/application_1586627890098_48586/container_e21_1586627890098_48586_02_000001/venv.zip/venv/lib/python3.6/site-packages/tensorflow_core/python/ops/math_grad.py:1424: where (from tensorflow.python.ops.array_ops) is deprecated and will be removed in a future version.
Instructions for updating:
Use tf.where in 2.0, which has the same broadcast rule as np.where
WARNING:tensorflow:From /dfs/1/yarn/nm/usercache/dw/appcache/application_1586627890098_48586/container_e21_1586627890098_48586_02_000001/analytics-zoo-bigdl_0.10.0-spark_2.4.3-0.9.0-SNAPSHOT-python-api.zip/zoo/tfpark/tf_optimizer.py:616: The name tf.keras.backend.get_session is deprecated. Please use tf.compat.v1.keras.backend.get_session instead.

creating: createRMSprop
creating: createSampleToMiniBatch
creating: createSampleToMiniBatch
creating: createZooKerasSparseCategoricalCrossEntropy
creating: createLoss
creating: createZooKerasSparseCategoricalAccuracy
WARNING:tensorflow:From /dfs/1/yarn/nm/usercache/dw/appcache/application_1586627890098_48586/container_e21_1586627890098_48586_02_000001/analytics-zoo-bigdl_0.10.0-spark_2.4.3-0.9.0-SNAPSHOT-python-api.zip/zoo/tfpark/tf_optimizer.py:157: The name tf.is_numeric_tensor is deprecated. Please use tf.debugging.is_numeric_tensor instead.

creating: createTFValidationMethod
creating: createTFValidationMethod
WARNING:tensorflow:From /dfs/1/yarn/nm/usercache/dw/appcache/application_1586627890098_48586/container_e21_1586627890098_48586_02_000001/analytics-zoo-bigdl_0.10.0-spark_2.4.3-0.9.0-SNAPSHOT-python-api.zip/zoo/tfpark/tf_optimizer.py:175: to_float (from tensorflow.python.ops.math_ops) is deprecated and will be removed in a future version.
Instructions for updating:
Use `tf.cast` instead.
WARNING:tensorflow:From /dfs/1/yarn/nm/usercache/dw/appcache/application_1586627890098_48586/container_e21_1586627890098_48586_02_000001/analytics-zoo-bigdl_0.10.0-spark_2.4.3-0.9.0-SNAPSHOT-python-api.zip/zoo/util/tf.py:37: The name tf.unsorted_segment_sum is deprecated. Please use tf.math.unsorted_segment_sum instead.

WARNING:tensorflow:From /dfs/1/yarn/nm/usercache/dw/appcache/application_1586627890098_48586/container_e21_1586627890098_48586_02_000001/analytics-zoo-bigdl_0.10.0-spark_2.4.3-0.9.0-SNAPSHOT-python-api.zip/zoo/tfpark/tf_optimizer.py:185: The name tf.GraphKeys is deprecated. Please use tf.compat.v1.GraphKeys instead.

WARNING:tensorflow:From /dfs/1/yarn/nm/usercache/dw/appcache/application_1586627890098_48586/container_e21_1586627890098_48586_02_000001/analytics-zoo-bigdl_0.10.0-spark_2.4.3-0.9.0-SNAPSHOT-python-api.zip/zoo/tfpark/tf_optimizer.py:199: The name tf.placeholder is deprecated. Please use tf.compat.v1.placeholder instead.

WARNING:tensorflow:From /dfs/1/yarn/nm/usercache/dw/appcache/application_1586627890098_48586/container_e21_1586627890098_48586_02_000001/analytics-zoo-bigdl_0.10.0-spark_2.4.3-0.9.0-SNAPSHOT-python-api.zip/zoo/tfpark/tf_optimizer.py:200: The name tf.assign is deprecated. Please use tf.compat.v1.assign instead.

WARNING:tensorflow:From /dfs/1/yarn/nm/usercache/dw/appcache/application_1586627890098_48586/container_e21_1586627890098_48586_02_000001/analytics-zoo-bigdl_0.10.0-spark_2.4.3-0.9.0-SNAPSHOT-python-api.zip/zoo/tfpark/tf_optimizer.py:239: The name tf.train.Saver is deprecated. Please use tf.compat.v1.train.Saver instead.

WARNING:tensorflow:From /dfs/1/yarn/nm/usercache/dw/appcache/application_1586627890098_48586/container_e21_1586627890098_48586_02_000001/analytics-zoo-bigdl_0.10.0-spark_2.4.3-0.9.0-SNAPSHOT-python-api.zip/zoo/tfpark/tf_optimizer.py:272: The name tf.tables_initializer is deprecated. Please use tf.compat.v1.tables_initializer instead.

WARNING:tensorflow:From /dfs/1/yarn/nm/usercache/dw/appcache/application_1586627890098_48586/container_e21_1586627890098_48586_02_000001/analytics-zoo-bigdl_0.10.0-spark_2.4.3-0.9.0-SNAPSHOT-python-api.zip/zoo/tfpark/tf_optimizer.py:281: The name tf.gfile.GFile is deprecated. Please use tf.io.gfile.GFile instead.

creating: createTFTrainingHelper
linux-x86_64/libiomp5.so
linux-x86_64/libmklml_intel.so
linux-x86_64/libtensorflow_framework-zoo.so
linux-x86_64/libtensorflow_jni.so
creating: createIdentityCriterion
creating: createSampleToMiniBatch
creating: createSampleToMiniBatch
creating: createEstimator
creating: createMaxEpoch
creating: createEveryEpoch
Traceback (most recent call last):
  File "ncf_dl_keras.py", line 175, in <module>
    model.fit(dataset,epochs=max_epoch,distributed=True)
  File "/dfs/1/yarn/nm/usercache/dw/appcache/application_1586627890098_48586/container_e21_1586627890098_48586_02_000001/analytics-zoo-bigdl_0.10.0-spark_2.4.3-0.9.0-SNAPSHOT-python-api.zip/zoo/tfpark/model.py", line 141, in fit
  File "/dfs/1/yarn/nm/usercache/dw/appcache/application_1586627890098_48586/container_e21_1586627890098_48586_02_000001/analytics-zoo-bigdl_0.10.0-spark_2.4.3-0.9.0-SNAPSHOT-python-api.zip/zoo/tfpark/model.py", line 164, in _fit_distributed
  File "/dfs/1/yarn/nm/usercache/dw/appcache/application_1586627890098_48586/container_e21_1586627890098_48586_02_000001/analytics-zoo-bigdl_0.10.0-spark_2.4.3-0.9.0-SNAPSHOT-python-api.zip/zoo/tfpark/tf_optimizer.py", line 703, in optimize
  File "/dfs/1/yarn/nm/usercache/dw/appcache/application_1586627890098_48586/container_e21_1586627890098_48586_02_000001/analytics-zoo-bigdl_0.10.0-spark_2.4.3-0.9.0-SNAPSHOT-python-api.zip/zoo/pipeline/estimator/estimator.py", line 103, in train_minibatch
  File "/dfs/1/yarn/nm/usercache/dw/appcache/application_1586627890098_48586/container_e21_1586627890098_48586_02_000001/analytics-zoo-bigdl_0.10.0-spark_2.4.3-0.9.0-SNAPSHOT-python-api.zip/zoo/common/utils.py", line 129, in callZooFunc
  File "/dfs/1/yarn/nm/usercache/dw/appcache/application_1586627890098_48586/container_e21_1586627890098_48586_02_000001/analytics-zoo-bigdl_0.10.0-spark_2.4.3-0.9.0-SNAPSHOT-python-api.zip/zoo/common/utils.py", line 123, in callZooFunc
  File "/opt/cloudera/parcels/SPARK2-2.4.0.cloudera2-1.cdh5.13.3.p0.1041012/lib/spark2/python/lib/py4j-0.10.7-src.zip/py4j/java_gateway.py", line 1257, in __call__
  File "/opt/cloudera/parcels/SPARK2-2.4.0.cloudera2-1.cdh5.13.3.p0.1041012/lib/spark2/python/lib/pyspark.zip/pyspark/sql/utils.py", line 63, in deco
  File "/opt/cloudera/parcels/SPARK2-2.4.0.cloudera2-1.cdh5.13.3.p0.1041012/lib/spark2/python/lib/py4j-0.10.7-src.zip/py4j/protocol.py", line 328, in get_return_value
py4j.protocol.Py4JJavaError: An error occurred while calling o66.estimatorTrainMiniBatch.
: java.lang.NullPointerException
	at scala.collection.mutable.ArrayOps$ofRef$.length$extension(ArrayOps.scala:192)
	at scala.collection.mutable.ArrayOps$ofRef.length(ArrayOps.scala:192)
	at scala.collection.SeqLike$class.size(SeqLike.scala:106)
	at scala.collection.mutable.ArrayOps$ofRef.size(ArrayOps.scala:186)
	at scala.collection.mutable.Builder$class.sizeHint(Builder.scala:69)
	at scala.collection.mutable.ArrayBuilder.sizeHint(ArrayBuilder.scala:22)
	at scala.collection.TraversableLike$class.builder$1(TraversableLike.scala:230)
	at scala.collection.TraversableLike$class.map(TraversableLike.scala:233)
	at scala.collection.mutable.ArrayOps$ofRef.map(ArrayOps.scala:186)
	at com.intel.analytics.zoo.pipeline.api.keras.models.InternalDistriOptimizer$.getLatestFile(Topology.scala:1531)
	at com.intel.analytics.zoo.pipeline.api.keras.models.InternalDistriOptimizer.train(Topology.scala:1229)
	at com.intel.analytics.zoo.pipeline.api.keras.models.InternalDistriOptimizer.train(Topology.scala:1382)
	at com.intel.analytics.zoo.pipeline.api.keras.models.InternalDistriOptimizer.train(Topology.scala:1070)
	at com.intel.analytics.zoo.pipeline.estimator.Estimator.train(Estimator.scala:152)
	at com.intel.analytics.zoo.pipeline.estimator.python.PythonEstimator.estimatorTrainMiniBatch(PythonEstimator.scala:106)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at py4j.reflection.MethodInvoker.invoke(MethodInvoker.java:244)
	at py4j.reflection.ReflectionEngine.invoke(ReflectionEngine.java:357)
	at py4j.Gateway.invoke(Gateway.java:282)
	at py4j.commands.AbstractCommand.invokeMethod(AbstractCommand.java:132)
	at py4j.commands.CallCommand.execute(CallCommand.java:79)
	at py4j.GatewayConnection.run(GatewayConnection.java:238)
	at java.lang.Thread.run(Thread.java:748)

**Comment by yangw1234** [comment]
@jack1981 this error is also unrelated to TFDataset, I have created another issue #2270

**Pull request merged**